### ISSUE TITLE
Grids to "data" to Fun?

### ISSUE THREAD

**Issue by gabrielgggg** [JuliaApproximation/ApproxFun.jl#605]
Hello! Sorry to post this here, I did not see a discussion section.
Can I use ApproxFun for this "workflow"? (The underlying function is expensive to evaluate.)

Say I have a function with 2 (or 3) arguments, which I would like to approximate over a square (or cube)...

I would like to
1) fix the order of the Chebyshev polynomial on each dimension (say 10 and 11, and maybe 6 for 3D case)
2) have the library provide me with the nodes where I must provide the value of the function
3) evaluate the function at those nodes and put together the "data" as a 2D (or 3D) array
4) feed the data (and the support/ranges and nodes?) to ApproxFun
5) use ApproxFun to evaluate the Chebyshev approximation anywhere inside the square/cube

I will need to do this iteratively, until some convergence criterion, so I'm trying to keep the amount of work needed at each iteration small and predictable.

tl;dr would like to use Chebyshev interpolation instead of b-splines or bi-linear etc.

Any advice is greatly appreciated!

**Comment by dlfivefifty** [comment]
You can do this in 2D, though ApproxFun's 2D support is based on Padua points, so its a vector of values, not a matrix:
```julia
using ApproxFun
    import ApproxFun: transform
f = (x,y) -> exp(x*y)

S = Chebyshev(0..1)^2

N = 10^2; p = points(S, N) # 105 padua points
f̃ = (xy) -> f(xy[1], xy[2])
F = f̃.(p)
F̌ = transform(S, F)  # 105 Chebyshev^2 coefficients ordered by polynomial degree 
f̌ = Fun(S, F̌)
f̌(0.1,0.2) ≈ f(0.1,0.2)
```
3D is not yet implemented.

Alternatively, you can use 1D ApproxFun to do what you asked, here's a quick hack:
```julia
S = Chebyshev(0..1)
p_x = points(S, 10)
p_y = points(S, 10)'
p_z = reshape(points(S, 10), 1, 1, 10)

f = (x,y,z) -> exp(x*y+z)

F = f.(p_x, p_y, p_z)

for k = 1:size(F,2),j=1:size(F,3)
    F[:,k,j] = transform(S,F[:,k,j])
end
for k = 1:size(F,1),j=1:size(F,3)
    F[k,:,j] = transform(S,F[k,:,j])
end
for k = 1:size(F,1),j=1:size(F,2)
    F[k,j,:] = transform(S,F[k,j,:])
end

f̌ = function(x,y,z)
    ret1 = Array{Float64}(undef, size(F,2), size(F,3))
    for j=1:size(F,2),l=1:size(F,3)
        ret1[j,l] = Fun(S,F[:,j,l])(x) 
    end
    ret2 = Array{Float64}(undef, size(F,3))
    for l=1:size(F,3)
        ret2[l] = Fun(S,ret1[:,l])(y) 
    end
    Fun(S, ret2)(z)
end

f̌(0.1,0.2,0.3) ≈ f(0.1,0.2,0.3)
```

**Comment by gabrielgggg** [comment]
Thank you very much! This looks perfect for me, assuming I understand correctly what the accented variables do.

**Comment by dlfivefifty** [comment]
Great! Note that there's some easy ways to make this more performant and less allocations, for example `P = plan_transform!(S, n)` will construct a pre-planned in-place transform.
